include/odp/api/spec/traffic_mngr.h

```diff
@@ -385,6 +408,31 @@ typedef struct {
 	 * the parameters of the threshold profile of any TM node or TM queue.
 	 */
 	odp_bool_t dynamic_threshold_update;
+
+	/** Indicates the packet priority modes supported by a TM systems on a platform.
+	 * A platform can support multiple packet priority modes. The actual mode,
+	 * a TM system runs with is defined by odp_tm_requirements_t.
+	 */
+	odp_bool_t pkt_prio_modes[ODP_TM_PKT_PRIO_MODE_MAX];
+
+	/** max_num_schedulers_per_node indicates maximum number of DWRR schedulers supported
+	 * by a tm node at any level. This capability is valid only when packet
+	 * priorirty mode is ODP_TM_PKT_PRIO_MODE_OVERRIDE.
+	 *
+	 * In ODP_TM_PKT_PRIO_MODE_OVERRIDE mode, since all pkts from a fan-in come with equal
+	 * priority, all fan-in nodes with equal priority form a group and a scheduler is
+	 * assigned to perform DWRR within the group. This capability indicates that at a given
+	 * node, max schedulers available which implies how many groups can be formed where DWRR
+	 * is required. Rest of the fan-in nodes should have unique priority.
+	 *
+	 * In ODP_TM_PKT_PRIO_MODE_GLOBAL mode, since packet priority is global,
+	 * every node should support as many number of schedulers as number of
+	 * priorities available. Hence this field is not applicable.
+	 *
+	 * This value can range from 1..ODP_TM_MAX_PRIORITIES as there can be
+	 * only as many number of WFQ groups as number of priorities.
+	 */
+	uint8_t max_num_schedulers_per_node;
```

Comment:
I wonder if we really need this capability field. The number of schedulers needed is no greater than the number of prioriries, so we could just rely on the maximum number of priorities for this instead of the separate capa field. odp_tm_level_capabilities_t already contains max_priority. This new capability field would seem to add value only in the case where there is only one fan-in with certain priorities (so that schedulers is not needed for them) and the max number of schedulers would would be enough only for the priorities that do have more than one fan-in. 

I do not quite understand the rationale for making this capability applicable only for the "override" mode. The "every node should support as many number of schedulers as number of priorities available" could as well be said about the "global" mode as well. In single-level TM system the user must not connect too many sets of queues of the different priorities to a node to not exceed the max scheduler count in the "override" mode. But the same can be done in the "global" mode too.

The text discusses DWRR (as if this capa would concern "only" DWRR schedulers and not other scheduler types such as (W)RR or WFQ), even though DWRR is not mentioned anywhere else in the API. The reference to DWRR should be removed, just "scheduler" should be enough.

Even though checkpatch no longer complains about lines longer than 80 characters but less than 100 characters, I think it would make sense to stick with the same max line length in this description text as the existing API text in this file. It just looks weird to have very different line lengths in the comment sections, sometimes for the same API element. This comment applies to other relevant changes too.

Reply:
>  This new capability field would seem to add value only in the case where there is only one fan-in with certain priorities (so that schedulers is not needed for them) and the max number of schedulers would would be enough only for the priorities that do have more than one fan-in.

Yes, that is the case in our hardware. We only have only one priority with multiple fan-in node and hence DWRR on them.
We support 10 priorities out of which only one priority can be having multiple fan-in's (those nodes overwriting priority value to 'X'). Rest of the 9 priorities can only have one fan-in nodes atmost.

> In single-level TM system the user must not connect too many sets of queues of the different priorities to a node to not exceed the max scheduler count in the "override" mode. But the same can be done in the "global" mode too.
But even in single-level TM system in global mode, multiple fan-ins can have same priorities. Say for example, if implementation supports 3 priorities (0..2) then there could be
SET1: Priority 0: (Fanin0, 1, 2.,..)
SET2: Priority 1: (fanin 3, 4, 5, 6, 7)
SET 3: Priority 2 : (fanin 9, 10, 11)

Now there are 12 queues connected to same Parent node with max three priorities. Now parent node has to run schedulers on each of the SET separately. In our case there can only one such SET where scheduler(DWRR or WRR or RR) is needed.
Rest of the SET's can have only one fan-in => no need to schedule.,

> The text discusses DWRR (as if this capa would concern "only" DWRR schedulers and not other scheduler types such as (W)RR or WFQ), even though DWRR is not mentioned anywhere else in the API. The reference to DWRR should be removed, just "scheduler" should be enough.
Will change it to scheduler.

>Even though checkpatch no longer complains about lines longer than 80 characters but less than 100 characters, I think it would make sense to stick with the same max line length in this description text as the existing API text in this file. It just looks weird to have very different line lengths in the comment sections, sometimes for the same API element. This comment applies to other relevant changes too.

Ack.

Reply:
> Yes, that is the case in our hardware. We can only have only one priority with multiple fan-in node and hence DWRR on them.

Ok. let's have this capa flag then.

> But even in single-level TM system in global mode, multiple fan-ins can have same priorities

Yes. And the exact same thing in non-global mode. I do not see why the capa should not apply to the global mode too. Let's say, there is a HW like yours (with limited number of schedulers) but without the priority overwriting. Then the capa would tell how I am allowed to configure things in global mode. IOW, an application can avoid TM config that exceeds the number of schedulers needed regardless of the priority mode. With single level TM the way to ensure that is the same but in multi-level config it is a but different. But to me, the capability itself does not seem mode specific at all.

Reply:
Ok, I'll remove the applicability string and for linux generic, set it as same as number of priorities supported.